### PR TITLE
#2651 - Fix hyperrectangle split with SVector

### LINKED ISSUE #2651: Splitting hyperrectangles with static vectors fails
This case is missing:

```julia
using LazySets, IntervalArithmetic

x = (0 .. 1) × (0 .. 1)
h = convert(Hyperrectangle, x)
split(h, [2, 2])

setindex!(::StaticArrays.SVector{2, Float64}, value, ::Int) is not defined.

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] setindex!(a::StaticArrays.SVector{2, Float64}, value::Float64, i::Int64)
   @ StaticArrays ~/.julia/packages/StaticArrays/NTbHj/src/indexing.jl:3
 [3] split(H::Hyperrectangle{Float64, StaticArrays.SVector{2, Float64}, StaticArrays.SVector{2, Float64}}, num_blocks::Vector{Int64})
   @ LazySets ~/.julia/dev/LazySets/src/Interfaces/AbstractHyperrectangle.jl:571
 [4] top-level scope
   @ In[4]:5
 [5] eval
   @ ./boot.jl:360 [inlined]
 [6] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
   @ Base ./loading.jl:1094

```

### PR DESCRIPTION
**Pull request by schillic**
Closes #2651.

### CODE REVIEW

File: src/Interfaces/AbstractHyperrectangle.jl [hunk 1]
```diff
@@ -555,7 +555,7 @@ A list of `Hyperrectangle`s.
 function split(H::AbstractHyperrectangle{N}, num_blocks::AbstractVector{Int}
               ) where {N}
     @assert length(num_blocks) == dim(H) "need number of blocks in each dimension"
-    radius = copy(radius_hyperrectangle(H))
+    radius = copy_vector(radius_hyperrectangle(H))
```
Comment (mforets) [inline src/Interfaces/AbstractHyperrectangle.jl:558]: This is OK but I suggest to use library functions when possible:

```julia
R = radius_hyperrectangle(H)
radius = similar(R)
copyto!(radius, R)
```
Returns an MArray if `R` is a static array.

File: src/Arrays/vector_operations.jl [hunk 2]
```diff
@@ -463,3 +463,35 @@ function find_unique_nonzero_entry(vector::AbstractVector{N}) where {N}
     end
     return res
 end
+
+function copy_vector(x::AbstractVector)
+    return copy(x)  # copy by default
+end
+
+function finalize_vector(x::AbstractVector)
```
Comment (mforets) [inline src/Arrays/vector_operations.jl:471]: Here I suggest to define

```julia
_similar_type(x::AbstractVector) = typeof(x)

# already exists
_similar_type(x::StaticArray) = StaticArrays.similar_type(x)
```
so we use it below as a constructor:

```julia
radius = _similar_type |> R |> radius
```

File: src/Interfaces/AbstractHyperrectangle.jl [hunk 3]
```diff
@@ -574,11 +574,13 @@ function split(H::AbstractHyperrectangle{N}, num_blocks::AbstractVector{Int}
             total_number *= m
         end
     end
+    radius = finalize_vector(radius)
 
     # create hyperrectangles for every combination of the center points
     result = Vector{Hyperrectangle{N, typeof(center(H)), typeof(radius)}}(undef, total_number)
     @inbounds for (i, center) in enumerate(product(centers...))
-        result[i] = Hyperrectangle(collect(center), copy(radius))
+        c = adapt_vector_type(collect(center), radius)
```
Comment (mforets) [inline src/Interfaces/AbstractHyperrectangle.jl:582]: Here we can also use `_similar_type` as a constructor:

```julia
c = _similar_type |> radius |> collect(center)
```

**Comment by schillic** [comment]
@mforets I do not fully follow. Feel free to apply the suggested changes!

**Comment by schillic** [comment]
Great, thanks, I agree with the changes. I just moved the code to another file because it was not specific to vectors anymore. If you approve we can merge.

**Review by mforets — APPROVED** [review]

**Pull request merged**